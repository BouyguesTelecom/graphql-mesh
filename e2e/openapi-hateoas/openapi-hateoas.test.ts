import { createTenv } from '@e2e/tenv';

describe('OpenAPI HATEOAS', () => {
  let server: any;

  beforeAll(done => {
    server = require('./server.js');
    done();
  });

  afterAll(done => {
    server.close(done);
  });

  const { compose, serve } = createTenv(__dirname);

  it('should compose', async () => {
    const { result } = await compose({ output: 'graphql' });
    expect(result).toMatchSnapshot();
  });

  it('should execute and follow HATEOAS links', async () => {
    const { output } = await compose({ output: 'graphql' });

    const { execute } = await serve({ supergraph: output });
    const queryResult = await execute({
      query: /* GraphQL */ `
        query GetProductsById {
          getProductById(id: 1) {
            id
            name
            supplierId
            supplier {
              id
              name
            }
          }
        }
      `,
    });

    expect(queryResult.errors).toBeFalsy();

    expect(queryResult.data).toEqual({
      getProductById: {
        id: 1,
        name: 'Laptop',
        supplierId: 11,
        supplier: {
          id: 11,
          name: 'Tech Supplier Inc.',
        },
      },
    });
  });
});
