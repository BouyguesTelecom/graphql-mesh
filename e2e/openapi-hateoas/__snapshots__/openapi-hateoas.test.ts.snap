// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI HATEOAS should compose 1`] = `
"schema @link(url: "https://specs.apollo.dev/link/v1.0") @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION) @link(url: "https://the-guild.dev/graphql/mesh/spec/v1.0", import: ["@httpOperation", "@transport", "@merge", "@extraSchemaDefinitionDirective"]) {
  query: Query
}

directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, usedOverridden: Boolean) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @join__graph(name: String!, url: String!) on ENUM_VALUE

directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE

directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true, isInterfaceObject: Boolean! = false) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA

directive @httpOperation(subgraph: String, path: String, operationSpecificHeaders: [[String]], httpMethod: HTTPMethod, isBinary: Boolean, requestBaseBody: ObjMap, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap, jsonApiFields: Boolean, queryStringOptions: ObjMap) repeatable on FIELD_DEFINITION

directive @transport(subgraph: String, kind: String, location: String, headers: [[String]], queryStringOptions: ObjMap, queryParams: [[String]]) repeatable on SCHEMA

directive @merge(subgraph: String, argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) repeatable on FIELD_DEFINITION

directive @extraSchemaDefinitionDirective(directives: _DirectiveExtensions) repeatable on OBJECT

directive @additionalField on FIELD_DEFINITION

scalar join__FieldSet

scalar link__Import

enum link__Purpose {
  """
  \`SECURITY\` features provide metadata necessary to securely resolve fields.
  """
  SECURITY
  """
  \`EXECUTION\` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

enum join__Graph {
  PRODUCTS @join__graph(name: "Products", url: "http://localhost:3000")
  SUPPLIERS @join__graph(name: "Suppliers", url: "http://localhost:3000")
}

scalar ObjMap @join__type(graph: PRODUCTS) @join__type(graph: SUPPLIERS)

scalar _DirectiveExtensions @join__type(graph: PRODUCTS) @join__type(graph: SUPPLIERS)

type Query @extraSchemaDefinitionDirective(directives: {transport: [{subgraph: "Products", kind: "rest", location: "http://localhost:3000"}]}) @extraSchemaDefinitionDirective(directives: {transport: [{subgraph: "Suppliers", kind: "rest", location: "http://localhost:3000"}]}) @join__type(graph: PRODUCTS) @join__type(graph: SUPPLIERS) {
  """Get a product by ID"""
  getProductById(
    """The product ID"""
    id: Int!
  ): Product @httpOperation(subgraph: "Products", path: "/products/{args.id}", operationSpecificHeaders: [["accept", "application/json"]], httpMethod: GET) @merge(subgraph: "Products", keyField: "id", keyArg: "id") @join__field(graph: PRODUCTS)
  """Get a supplier by ID"""
  getSupplierById(
    """The supplier ID"""
    id: Int!
  ): Supplier @httpOperation(subgraph: "Suppliers", path: "/suppliers/{args.id}", operationSpecificHeaders: [["accept", "application/json"]], httpMethod: GET) @merge(subgraph: "Suppliers", keyField: "id", keyArg: "id") @join__field(graph: SUPPLIERS)
}

type Product @join__type(graph: PRODUCTS, key: "id") {
  id: Int
  _links: ProductLinks
  name: String
  price: Float
  supplierId: Int
  self: Product @resolveTo(sourceName: "Products", sourceTypeName: "Query", sourceFieldName: "getProductById", requiredSelectionSet: "{ id }", sourceArgs: {id: "{root.id}"}) @additionalField
  supplier: Supplier @resolveTo(sourceName: "Suppliers", sourceTypeName: "Query", sourceFieldName: "getSupplierById", requiredSelectionSet: "{ supplierId }", sourceArgs: {id: "{root.supplierId}"}) @additionalField
}

type ProductLinks @join__type(graph: PRODUCTS) {
  self: Link!
  supplier: Link!
}

type Link @join__type(graph: PRODUCTS) {
  href: String!
}

type Supplier @join__type(graph: SUPPLIERS, key: "id") {
  id: Int
  name: String
  address: String
}

enum HTTPMethod @join__type(graph: PRODUCTS) @join__type(graph: SUPPLIERS) {
  GET @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  HEAD @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  POST @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  PUT @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  DELETE @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  CONNECT @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  OPTIONS @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  TRACE @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
  PATCH @join__enumValue(graph: PRODUCTS) @join__enumValue(graph: SUPPLIERS)
}"
`;
